tracing-subscriber/src/registry/mod.rs

```diff
@@ -45,3 +49,104 @@ pub trait LookupMetadata {
         self.metadata(id).is_some()
     }
 }
+
+pub trait LookupSpan<'a>
+where
+    Self: std::marker::Sized,
+{
+    type Data: SpanData<'a>;
+    fn span_data(&'a self, id: &Id) -> Option<Self::Data>;
+
+    fn span(&'a self, id: &Id) -> Option<SpanRef<'a, Self>> {
+        let data = self.span_data(id)?;
+        Some(SpanRef {
+            registry: self,
+            data,
+        })
+    }
+}
+
+pub trait SpanData<'a> {
+    type Children: Iterator<Item = &'a Id>;
+    type Follows: Iterator<Item = &'a Id>;
+
+    fn id(&self) -> Id;
+    fn metadata(&self) -> &'static Metadata<'static>;
+    fn parent(&self) -> Option<&Id>;
+    fn children(&'a self) -> Self::Children;
+    fn follows_from(&'a self) -> Self::Follows;
+}
+
+// XXX(eliza): should this have a SpanData bound? The expectation is that there
+// would be add'l impls for `T: LookupSpan where T::Span: Extensions`...
+//
+// XXX(eliza): also, consider having `.extensions`/`.extensions_mut` methods to
+// get the extensions, so we can control read locking vs write locking?
+pub trait Extensions {
+    fn get<T: Any>(&self) -> Option<&T>;
+    fn get_mut<T: Any>(&mut self) -> Option<&mut T>;
+    fn insert<T: Any>(&mut self, t: T) -> Option<T>;
+    fn remove<T: Any>(&mut self) -> Option<T>;
+}
+
+#[derive(Debug)]
+pub struct SpanRef<'a, R: LookupSpan<'a>> {
+    registry: &'a R,
+    data: R::Data,
+}
+
+#[derive(Debug)]
+pub struct Parents<'a, R> {
+    registry: &'a R,
+    next: Option<Id>,
+}
+
+impl<'a, R> SpanRef<'a, R>
+where
+    R: LookupSpan<'a>,
+{
+    pub fn id(&self) -> Id {
+        self.data.id()
+    }
+
+    pub fn parent_id(&self) -> Option<&Id> {
+        self.data.parent()
+    }
+
+    pub fn parent(&self) -> Option<Self> {
+        let id = self.data.parent()?;
+        let data = self.registry.span_data(id)?;
+        Some(Self {
+            registry: self.registry,
+            data,
+        })
+    }
+
+    pub fn parents(&'a self) -> Parents<'_, R> {
+        Parents {
+            registry: self.registry,
+            next: self.parent().map(|parent| parent.id()),
+        }
+    }
+
+    pub fn child_ids(&'a self) -> <R::Data as SpanData<'a>>::Children {
+        self.data.children()
+    }
+}
+
+impl<'a, R> Iterator for Parents<'a, R>
+where
+    R: LookupSpan<'a>,
+{
+    type Item = SpanRef<'a, R>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let id = self.next.take()?;
+        let span = self.registry.span(&id)?;
+        if let Some(parent) = span.parent() {
+            self.next = Some(parent.id());
+            Some(span)
+        } else {
+            None
+        }
```

Comment:
This is incorrect, it returns `None` when the span we are about to return has no parent. This means that any spans that are immediate children of the root span will be ignored.

What about:
```rust
self.next = span.parent().map(|parent| parent.id());
Some(span)
```
or something similar?

Reply:
Ah, that makes sense. Just applied that fix.